pandas/tests/reshape/test_merge.py

```diff
@@ -1208,11 +1204,179 @@ def f():
+                     TripPurp=['hbw', 'nhb', 'hbo', np.nan, 'nhb',
+                               'hbw', np.nan, np.nan],
+                     LinkType=['a', 'a', 'c', 'b', np.nan, 'a', 'b', 'a'],
+                     Trips=[1987, 3647, 2470, np.nan,
+                            4296, 4444, np.nan, np.nan],
+                     Distance=[100, 80, 90, 80, np.nan, 75, 35, 55]),
+                columns=['Origin', 'Destination', 'Period', 'TripPurp',
+                         'LinkType', 'Trips', 'Distance'])
+            .set_index(['Origin', 'Destination', 'Period',
+                        'TripPurp', 'LinkType']))
+
+        result = matrix.join(distances, how='outer')
+        assert_frame_equal(result, expected)
+
+        # Non-unique resulting index
+        distances2 = (
+            pd.DataFrame(
+                dict(Origin=[1, 1, 2],
+                     Destination=[1, 1, 1],
+                     Period=['AM', 'AM', 'PM'],
+                     LinkType=['a', 'b', 'a'],
+                     Distance=[100, 110, 120]),
+                columns=['Origin', 'Destination', 'Period',
+                         'LinkType', 'Distance'])
+            .set_index(['Origin', 'Destination', 'Period', 'LinkType']))
+
+        def f():
+            matrix.join(distances2, how='left')
+        pytest.raises(TypeError, f)
+
+        # No-overlapping level names
+        distances2 = (
+            pd.DataFrame(
+                dict(Orig=[1, 1, 2, 2, 3, 3, 5],
+                     Dest=[1, 2, 1, 2, 1, 2, 6],
+                     Per=['AM', 'PM', 'IP', 'AM', 'OP', 'IP', 'AM'],
+                     LinkTyp=['a', 'a', 'c', 'b', 'a', 'b', 'a'],
+                     Dist=[100, 80, 90, 80, 75, 35, 55]),
+                columns=['Orig', 'Dest', 'Per', 'LinkTyp', 'Dist'])
+            .set_index(['Orig', 'Dest', 'Per', 'LinkTyp']))
 
         def f():
-            household.join(log_return, how='outer')
-        pytest.raises(NotImplementedError, f)
+            matrix.join(distances2, how='left')
+        pytest.raises(ValueError, f)
+
+        # Empty Level
+        distances2 = (
+            pd.DataFrame(
+                dict(Origin=[1, 1, 2, 2, 3, 3, 5],
+                     Destination=[1, 2, 1, 2, 1, 2, 6],
+                     Period=[np.nan] * 7,
+                     LinkType=['a', 'a', 'c', 'b', 'a', 'b', 'a'],
+                     Distance=[100, 80, 90, 80, 75, 35, 55]),
+                columns=['Origin', 'Destination', 'Period',
+                         'LinkType', 'Distance'])
+            .set_index(['Origin', 'Destination', 'Period', 'LinkType']))
+
+        expected = (
+            pd.DataFrame(
+                dict(Origin=[1, 1, 2, 2, 3],
+                     Destination=[1, 2, 1, 3, 1],
+                     Period=['AM', 'PM', 'IP', 'AM', 'OP'],
+                     TripPurp=['hbw', 'nhb', 'hbo', 'nhb', 'hbw'],
+                     Trips=[1987, 3647, 2470, 4296, 4444],
```

Comment:
in fact why don't you move all of these added tests to a new TestMultiMulti class for testing (and other relevant tests if any).